### ISSUE TITLE
Noob question about actions, why not just to batch all mutations into setImmediate?

### ISSUE THREAD

**Issue by roginvs** [mobxjs/mobx#1523]
I am starting to learn mobx, and I do not understand why mobx invented "actions" entity. Will it be easier just to batch all changes into next tick inside setImmediate? That will automatically make all sync state changes act in the same way as @action do now. Is there any profit of triggering observers right just after action finished instead of inside next tick?

**Comment by urugator** [comment]
Few ideas:

`setImmediate` is not part of the standard (afaik it's implemented only by node and IE)

Batching is opt-in concept.

https://hackernoon.com/the-fundamental-principles-behind-mobx-7a725f71f3e8
> So far UI libraries have always taken the easy way out when it comes to scheduling derivations: Mark derivations as dirty, and re-run them on the next tick after all the state has been updated. This is simple and straight forward. It is a fine approach if your only concern is updating the DOM. The DOM usually lags a bit ‘behind’ and we hardly read data from it programmatically. So temporary staleness is not really an issue. Yet temporary staleness kills the general applicability of a reactive library.

**Comment by roginvs** [comment]
As for me, it will be good to keep `@action` as it is, but every mutation outside `action` instead of throwing an error should start a transaction which will automatically end (commit) in next tick.
PS: `setTimeout(fn, 0)` should work similar to `setImmediate`

**Comment by urugator** [comment]
`setTimeout` is not synchronous (and potentially "slow", by specifiaction it always waits at least 4ms). 
Check the link why mobx is synchronous by design.

**Comment by roginvs** [comment]
But React somehow works that way, is not it? And no one complains that React batches state updates.
Also, `@computed` values are updated even inside `@action`, so I do not understand any potential problem with this. If I get it right, then almost every observer will be a React component.

**Comment by urugator** [comment]
Also one thing I forgot to mention is that action prevents from unintended subscriptions.

**Comment by roginvs** [comment]
> So anytime you read a state outside of action/transaction (even outside of reaction) it's up to date, consistent and safe to read...
yes, right, that concept I do agree with. I am talking about to batch all updates inside all React components.
Actually I want to do `<input value={appState.name} onChange={e => appState.name = e.target.value}/>` without any warnings, meaning if I do not made an `action`, then I do not actually care how immediately will it update

**Comment by urugator** [comment]
As mentioned React batches updates only when they are performed from event listener callback. It's the only situation when React knows where the batch should begin/end. Afaik it doesn't rely on `setTimeout` or anything like that...
However React exposes `ReactDOM.unstable_batchedUpdates(callback)` (it's exactly the same piece of code which wraps event listeners). It allows to manually delimit the "batch border", similary to Mobx's `action`.
Mobx uses this function to prevent re-rending each notified component individually, roughly:  
```javascript
// At the end of mobx transaction:
ReactDOM.unstable_batchedUpdates(() => {
  runScheduledReactions() // invokes forceUpdate on multiple individual components     
})
```
Notice that if we would use `setTimeout` to schedule reactions, we wouldn't be able to pass them to React's batchedUpdates... So integration with React is another reason why reactions must run synchronously (the components must be notified synchronously, event though we don't care about actual DOM update)...
